idle.go

```diff
@@ -0,0 +1,365 @@
+func (i *atomicIdlenessManager) enterIdleMode() bool {
+	i.idleMu.Lock()
+	defer i.idleMu.Unlock()
+
+	// Compare-and-swap the active calls count with an old value of `0` and a
+	// new value of `minInt32`. If this succeeds, it means that the active calls
+	// count is still zero, i.e no RPCs are ongoing since we last checked the
+	// count in the timer callback. If this fails, it means that RPCs have
+	// started since we last checked the count in the timer callback, and
+	// therefore we should not ask the ClientConn to enter idle mode.
+	if !atomic.CompareAndSwapInt32(&i.activeCallsCount, 0, math.MinInt32) {
+		return false
+	}
+	// It is possible that one or more RPCs started and finished since the last
+	// time we checked the calls count and activity in the timer callback. In
+	// this case, the calls count would be zero and the above compare-and-swap
+	// operation would have succeeded. The `activeSinceLastTimerCheck` field is
+	// set to false only from the timer callback and is set to true in
+	// onCallBegin. Therefore, if one or more RPCs started and finished before
+	// we got here, this field would be set to true. And in this case, we don't
+	// want to enter idle mode.
+	if active := atomic.LoadInt32(&i.activeSinceLastTimerCheck); active == 1 {
+		return false
+	}
+	if err := i.enforcer.enterIdleMode(); err != nil {
+		logger.Errorf("Failed to enter idle mode: %v", err)
+		return false
+	}
+	// We are in enter idle mode now. Set the active calls count appropriately.
+	atomic.StoreInt32(&i.activeCallsCount, math.MinInt32)
+	return true
+}
+
+// exitIdleMode instructs the ClientConn to exit idle mode.
+//
+// Holds idleMu which ensures mutual exclusion with enterIdleMode.
+func (i *atomicIdlenessManager) exitIdleMode() error {
+	i.idleMu.Lock()
+	defer i.idleMu.Unlock()
+
+	// When the channel enters idle mode, the active calls count is set to
+	// math.MinInt32. RPCs that start when the channel is in idle mode increment
+	// the active calls count in onCallBegin. If there are multiple RPCs
+	// competing to get the channel out of idle mode, the first one to grab the
+	// lock here gets to do so and sets the active calls count back to 1. So, if
+	// the count is not negative here, we know that someone else won the race
+	// and moved the channel out of idle mode, and we have nothing to do here.
+	if count := atomic.LoadInt32(&i.activeCallsCount); count >= 0 {
+		atomic.AddInt32(&i.activeCallsCount, 1)
+		return nil
+	}
+
+	// The first one to grab the lock will get here to instruct the channel to
+	// move out of idle mode.
+	if err := i.enforcer.exitIdleMode(); err != nil {
+		return status.Errorf(codes.Internal, "grpc: ClientConn failed to exit idle mode: %v", err)
+	}
+
+	// Reset the calls count to 1 and reset the timer to fire after a duration
+	// of the configured idle timeout.
+	atomic.StoreInt32(&i.activeCallsCount, 1)
+	i.timer.Store(timeAfterFunc(time.Duration(i.timeout), i.handleIdleTimeout))
```

Comment:
Oh BTW I think you may want to put all the timer accesses inside the mutex.  The only one that happens on the RPC path is when exiting idle and we're already holding the mutex anyway.  Doing it with atomics means you could leak the timer on close() if a new timer is created immediately after stopping the old one.

Reply:
Done.